src/Storages/StorageS3.cpp

```diff
@@ -438,7 +485,15 @@ StorageS3Source::KeyWithInfoPtr StorageS3Source::DisclosedGlobIterator::next(siz
 
 size_t StorageS3Source::DisclosedGlobIterator::estimatedKeysCount()
 {
-    return pimpl->objectsCount();
+    if (pimpl->hasMore())
+    {
+        /// 1000 files were listed, and we cannot make any estimation of _how many more_ there are (because we list bucket lazily);
+        /// If there are more objects in the bucket, limiting the number of streams is the last thing we may want to do
+        /// as it would lead to serious (up to <max_threads> times) reading performance degradation.
```

Comment:
```suggestion
        /// as it would lead to serious slow down of the execution, since objects are going 
        // to be fetched sequentially rather than in-parallel with up to <max_threads> times.
```

This is not a performance degradation, more like suboptimal non-parallel execution. But this is a minor nitpick